### ISSUE TITLE
Try out PAL?

### ISSUE THREAD

**Issue by nilscrm** [nilscrm/stackelberg-ml#25]
I've been thinking whether it might be worth it to try out PAL. This would make sense especially when we have some inner loop training for the model. My reasoning is the following:
1. In a Stackelberg game, the leader always has a strict advantage (can achieve at least as much value as if both acted at the same time) and our main objective should be to maximies the reward not the accuracy of the model
2. With complete pre-training of the model, we would be somewhat in model based RL. However, when we do some fine tuning steps in the inner loop, we only train our model for the states our policy actually visits.
3. I think there are some nice theoretical properties of PAL. Currently with MAL, we can think of the task of the model as "knowing what the optimal policy for a model is, how can I design a model in which I can predict the optimal policy well". This naturally leads to undesired behaviour when the model can design very uninteresting models. Now look at the objective of the model in PAL. It would be something like "Given a policy, what is a best model to predict the trajectories of that policy". This has 2 very nice properties. First, we are actually incentivised to predict the state transitions that the policy takes accurately. Second, this still allows the model to predict the world inaccurately where the policy is not going.

**Comment by YanickZengaffinen** [comment]
Agree, this makes sense in an inner-outer loop setting. Might also be interesting to look at what params GT-MBRL ended up using (bc this kind of suggests that the goal of PAL is more aligned with the goal of MBRL generally)

**Comment by nilscrm** [comment]
Interesting that you can see that PAL is actually doing better in there experiments as well.

<img width="1146" alt="image" src="https://github.com/nilscrm/stackelberg-ml/assets/38761406/9576a65d-e19d-45b6-82bc-2c3d53bf8b44">

**Comment by YanickZengaffinen** [comment]
Implementation Specification: 
- add train_contextualized_PAL
- use SGD with kldivloss (or CE loss) to pretrain model (now we don't have the problem anymore of showing policy queries to the model training)
- use normal PPO for leader training
- support inner loop before doing training step (1. make copy of current oracle weights, 2. train follower to convergence, 3. train leader, 4. reset oracle weights to the copied ones)

**Comment by YanickZengaffinen** [comment]
PAL implemented and working
